### PR TITLE
Domain change: Premium domains

### LINKED ISSUE #8356: Display premium domains in the suggestion list

### PR DESCRIPTION
**Pull request by 0nko**
Fixes #8356, a subtask of #8238.

This PR implements the premium domain search & price lookup. 2 relatively bigger changes were introduced in the domain search ViewModel:

- `DomainPickerViewModel` has been renamed to `DomainSuggestionsViewModel` and moved to the `ui.common` package, so it can be reused and extended with specific implementation details (e.g. specific analytics tracking, help origin tag and domain selection handling, which are different for store creation and domain change picker)
- The original implementation used the `StateFlow` for domain search TextField state management, which resulted in weird/incorrect behaviour when I added domain price lookup. In fact, I learned that for this very reason [the use of `StateFlow` has been discouraged](https://developer.android.com/jetpack/compose/text#effective-state). So changed it the suggested Compose `MutableState`.

**Premium domain search:**
<video src="https://user-images.githubusercontent.com/1522856/220200319-61357463-ba95-4472-b86b-416dfe4c44e6.webm" />

**Free domain search:**
<video src="https://user-images.githubusercontent.com/1522856/220200480-2610c94a-fc11-49ea-9414-09f05d481aa7.webm" />

**To test:**
1. Start the app and go to the site picker
2. Trigger the store creation and navigate to the domain picker
3. Smoke test the domain search and verify it behaves as expected

1. Log in to a site with Business plan that is eligible for a free domain
2. Go to Settings -> Domains
3. Tap on Claim Domain CTA
4. Type a keyword that will produce a premium domain (for example "blablabla")
5. Notice the search result shows a domain that has a high price and the domain is not free

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/domain/CurrentDomainsFragment.kt [hunk 1]
```diff
@@ -50,7 +50,7 @@ class CurrentDomainsFragment : BaseFragment() {
                 is ShowMoreAboutDomains -> ChromeCustomTabUtils.launchUrl(requireContext(), event.url)
                 is NavigateToDomainSearch -> findNavController().navigate(
                     CurrentDomainsFragmentDirections
-                        .actionCurrentDomainFragmentToDomainSearchFragment(event.hasFreeCredits, false)
+                        .actionCurrentDomainFragmentToDomainSearchFragment(event.hasFreeCredits, false, "")
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/domain/CurrentDomainsFragment.kt:53]: Np, could we use named params here? At a first glance I couldn't tell what the `false` and `""` values were

**Comment by JorgeMucientes** [comment]
Outstanding job @0nko 🥇.  It works like a charm! Ready to merge. Let's see if we can get FluxC issues resolved soon in order to merge this PRs

**Review by JorgeMucientes — APPROVED** [review]

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        )
+    }.asLiveData()
+
+    init {
+        viewModelScope.launch {
+            if (!searchOnlyFreeDomains) {
+                domainSuggestionsRepository.fetchProducts()
+            }
+
+            snapshotFlow { domainQuery }
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt:81]: Nice! TIL about this `snapshotFlow` and the bad usage I was previously doing of stateFlow to represent text changes. Good call on refactoring this 👍🏼

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/domain/CurrentDomainsFragment.kt (hunk 1 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/domain/CurrentDomainsFragment.kt:53]: Never mind about this! I saw you added named params in the "next" PR.

Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/domain/CurrentDomainsFragment.kt:53]: Yes, sorry about that, I keep fixing stuff like that with the PRs getting queued up.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt (hunk 2 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt:81]: I didn't know about it either until the TextField started acting weird and I looked it up 😄

**Comment by 0nko** [comment]
🎉 Thanks for the review, Jorge!

**Pull request merged**